### ISSUE TITLE
MQTT does not re-connect if disconnected for a while/new host

### ISSUE THREAD

**Issue by lyuzashi** [lyuzashi/SparkGap#2]
**Replication:**
* Shut down MQTT server
* Wait (not sure how long)
* Start MQTT server
* Observe SparkGap LED continues to flash, never connects

**Investigation:**
* Using the callback-like loops, it is easily possible that a stack overflow is being created
* Unlikely related to a potential bug with finding a new host, since starting the original host did not trigger a stable connect
* Possible that service advertisement did not happen correctly

**Proposal:**
* Re-write callbacks into an "on next loop" system to prevent stack overflow

**Comment by lyuzashi** [comment]
Using "on next loop" does help by reducing the stack size, but still some crashes occur while attempting to find a new MQTT server to connect to. 

**Next to try:**
Adding a delay to dns find

```

Decoding stack results
0x402024bc: operator new(unsigned int) at SparkGap/utils/esp8266/cores/esp8266/abi.cpp line 30
0x40203fc8: std::vector  , std::allocator   > >::_M_emplace_back_aux   const&>(std::function  const&) at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/bits/vector.tcc line 415
0x402022b4: esp_yield() at SparkGap/utils/esp8266/cores/esp8266/core_esp8266_main.cpp line 87
0x4020137a: delay at SparkGap/utils/esp8266/cores/esp8266/core_esp8266_wiring.c line 54
0x40204040: std::vector  , std::allocator   > >::push_back(std::function  const&) at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/bits/stl_vector.h line 915
0x4020405b: LoopQueue::onLoop(std::function ) at SparkGap/src/loop-queue.cpp line 11
0x40204346: DNSSD::updateState(int) at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/functional line 2174
0x4020425c: std::_Function_base::_Base_manager ::_M_manager(std::_Any_data &, const std::_Any_data &, std::_Manager_operation) at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/functional line 1931
0x4020423c: std::_Function_handler ::_M_invoke(const std::_Any_data &) at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/functional line 2069
0x4020261c: dnssdState(int) at SparkGap/src/SparkGap.ino line 56
0x4020807c: std::_Function_base::_Base_manager ::_M_manager(std::_Any_data&, std::_Any_data const&, std::_Manager_operation) at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/functional line 1934
0x40208068: std::_Function_handler ::_M_invoke(std::_Any_data const&, int) at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/functional line 2069
0x4020261c: dnssdState(int) at SparkGap/src/SparkGap.ino line 56
0x40208150: std::_Function_handler    (LED*)> >::_M_invoke(std::_Any_data const&) at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/functional line 2069
0x4020391c: LED::blink() at SparkGap/src/led.cpp line 49
0x40204401: DNSSD::find() at SparkGap/src/dnssd.cpp line 26
0x40202632: dnssdState(int) at SparkGap/src/SparkGap.ino line 61
0x40205754: ESP8266WiFiSTAClass::status() at SparkGap/utils/esp8266/libraries/ESP8266WiFi/src/ESP8266WiFiSTA.cpp line 483
0x40202519: wpsState(int) at SparkGap/src/SparkGap.ino line 50
0x40208074: std::_Function_handler ::_M_invoke(std::_Any_data const&, int) at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/functional line 2073
0x40204254: std::_Function_handler ::_M_invoke(const std::_Any_data &) at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/functional line 2073
0x40203ef2: std::function ::operator()() const at SparkGap/utils/esp8266/tools/xtensa-lx106-elf/xtensa-lx106-elf/include/c++/4.8.2/functional line 2465
0x40203f22: LoopQueue::loop() at SparkGap/src/loop-queue.cpp line 18
0x40202798: loop() at SparkGap/src/SparkGap.ino line 125
0x40202300: loop_wrapper() at SparkGap/utils/esp8266/cores/esp8266/core_esp8266_main.cpp line 124

```

**Comment by lyuzashi** [comment]
Another option to try is disabling the blink timer, since that appears in the stack and could possibly be interfering with the mDNS wait

**Comment by lyuzashi** [comment]
Seems to be solved with delayed loop retry 83bfc20a48c16291d09cea1888d5c0acda821241
